clang/lib/Gold/GoldDeclarationBuilder.cpp

```diff
@@ -0,0 +1,814 @@
+  // are template template parameters, in which case they should be specified
+  // differently.
+  EnableTemplateParameters = false;
+  RequireTypeForVariable = true;
+  EnableNestedNameSpecifiers = false;
+  RequireAliasTypes = false;
+  RequireTypeForFunctions = false;
+  RequiresDeclOrError = true;
+  return makeDeclarator(S);
+}
+
+Declarator *DeclarationBuilder::handleFunctionScope(const Syntax *S) {
+  EnableFunctions = true;
+  EnableNamespaceDecl = false;
+  EnableTags = true;
+  EnableAliases = true;
+  EnableTemplateParameters = false;
+  RequireTypeForVariable = false;
+  EnableNestedNameSpecifiers = false;
+  RequireAliasTypes = true;
+  RequireTypeForFunctions = true;
+  return makeDeclarator(S);
+}
+
+Declarator *DeclarationBuilder::handleBlockScope(const Syntax *S) {
+  EnableFunctions = true;
+  EnableNamespaceDecl = false;
+  EnableTags = true;
+  EnableAliases = true;
+  EnableTemplateParameters = false;
+  RequireTypeForVariable = false;
+  EnableNestedNameSpecifiers = false;
+  RequireAliasTypes = true;
+  RequireTypeForFunctions = true;
+  return makeDeclarator(S);
+}
+
+Declarator *DeclarationBuilder::handleClassScope(const Syntax *S) {
+  EnableFunctions = true;
+  EnableNamespaceDecl = false;
+  EnableTags = true;
+  EnableAliases = true;
+  EnableTemplateParameters = true;
+  RequireTypeForVariable = true;
+  EnableNestedNameSpecifiers = false;
+  RequireAliasTypes = false;
+  RequireTypeForFunctions = false;
+  RequiresDeclOrError = true;
+  return makeDeclarator(S);
+}
+
+Declarator *DeclarationBuilder::handleControlScope(const Syntax *S) {
+  EnableFunctions = false;
+  EnableNamespaceDecl = false;
+  EnableTags = false;
+  EnableAliases = false;
+  RequireTypeForVariable = false;
+  EnableTemplateParameters = false;
+  EnableNestedNameSpecifiers = false;
+  RequireAliasTypes = false;
+  RequireTypeForFunctions = false;
+  return makeDeclarator(S);
+}
+
+Declarator *DeclarationBuilder::handleEnumScope(const Syntax *S) {
+  // This is done 100% Seperate from all other declarations, because of how
+  // limited enum declarations actually are.
+  if (const auto *NameOnly = dyn_cast<AtomSyntax>(S)) {
+    return handleIdentifier(NameOnly, nullptr);
+  } else if (const auto *Call = dyn_cast<CallSyntax>(S)) {
+    EnableFunctions = false;
+    EnableNamespaceDecl = false;
+    EnableTags = false;
+    EnableAliases = false;
+    RequireTypeForVariable = false;
+    EnableTemplateParameters = false;
+    EnableNestedNameSpecifiers = false;
+    RequireAliasTypes = false;
+    RequireTypeForFunctions = false;
+    IsInsideEnum = true;
+    return makeDeclarator(Call);
+  } else if (const ErrorSyntax *Es = dyn_cast<ErrorSyntax>(S)) {
+    return handleErrorSyntax(Es, nullptr);
+  } else {
+    SemaRef.Diags.Report(S->getLoc(),
+                         clang::diag::err_invalid_declaration);
+    return nullptr;
+  }
+}
+
+static bool isParameterSyntax(const Syntax *S) {
+  if (const auto * Call = dyn_cast<CallSyntax>(S)) {
```

Comment:
This function is basically unreadable. There's a lot of unnecessary casting going on. It could be fixed like so:
```
FusedOpKind Op = getFusedOpKind(SemaRef, Call);
if (Op == FOK_Colon) {
  return true;
} else if (Op == FOK_Equals) {
  Syntax *Arg = dyn_cast<CallSyntax>(Call)->getArgument(0);
  if (!Arg)
    return false;
  if (getFusedOpKind(SemaRef, dyn_cast<CallSyntax>(Arg)) == FOK_Colon)
    return true;
  }

  return false;
}
```

The reason Clang originally started using the 80 character limit was to basically prevent wild cast strings and 8-if-statements-deep nesting, it's trying to force you to make sparing use of RTTI.

I would ask that you prefer just creating a variable any time you have to split an if-statement on the equals.